Raspberry Pi OnnxRuntime error: call of overloaded ‘abs(__gnu_cxx::__alloc_traits<std::allocator<float> >::value_type)’ is ambiguous ASSERT_TRUE(std::abs(values_y[i] - f[i]) < 1e-6)

**Describe the bug**
I'm following the steps given [here](https://github.com/microsoft/onnxruntime/blob/master/dockerfiles/README.md#arm-32v7) to build onnxruntime on Raspberry Pi. The installation fails with the below error message. This issue has been discussed in #4393 but the platform was Linux and the answer couldn't clear the issue in my case. 



**System information**
- OS Platform and Distribution: Raspbian Stretch
- ONNX Runtime installed from (source or binary): Source
- ONNX Runtime version: master branch
- Python version: 3.5.3
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source): 6.3.0

**To Reproduce**
At step: RUN ./build.sh --use_openmp ${BUILDARGS} --build_shared_lib from the Dockerfile.arm32v7 file
The error output is: 
[100%] Building CXX object CMakeFiles/onnxruntime_global_thread_pools_test.dir/code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc.o

In file included from /code/onnxruntime/cmake/external/googletest/googletest/include/gtest/gtest.h:62:0,
                 from /code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc:14:
/code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc: In instantiation of ‘void RunSession(OrtAllocator&, Ort::Session&, std::vector<Input>&, const char*, const std::vector<long long int>&, const std::vector<T>&, Ort::Value*) [with OutT = float; OrtAllocator = OrtAllocator]’:
/code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc:108:19:   required from ‘void TestInference(Ort::Session&, std::vector<Input>&, const char*, const std::vector<long long int>&, const std::vector<OutT>&) [with T = const char*; OutT = float]’
/code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc:157:109:   required from here
**/code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc:61:20: error: call of overloaded ‘abs(__gnu_cxx::__alloc_traits<std::allocator<float> >::value_type)’ is ambiguous
     ASSERT_TRUE(std::abs(values_y[i] - f[i]) < 1e-6);**
                 ~~~^~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/cstdlib:75:0,
                 from /usr/include/c++/6/ext/string_conversions.h:41,
                 from /usr/include/c++/6/bits/basic_string.h:5417,
                 from /usr/include/c++/6/string:52,
                 from /usr/include/c++/6/stdexcept:39,
                 from /usr/include/c++/6/array:39,
                 from /usr/include/c++/6/tuple:39,
                 from /usr/include/c++/6/functional:55,
                 from /usr/include/c++/6/memory:79,
                 from /code/onnxruntime/include/onnxruntime/core/common/make_unique.h:28,
                 from /code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc:4:
/usr/include/stdlib.h:735:12: note: candidate: int abs(int)
 extern int abs (int __x) __THROW __attribute__ ((__const__)) __wur;
            ^~~
In file included from /usr/include/c++/6/ext/string_conversions.h:41:0,
                 from /usr/include/c++/6/bits/basic_string.h:5417,
                 from /usr/include/c++/6/string:52,
                 from /usr/include/c++/6/stdexcept:39,
                 from /usr/include/c++/6/array:39,
                 from /usr/include/c++/6/tuple:39,
                 from /usr/include/c++/6/functional:55,
                 from /usr/include/c++/6/memory:79,
                 from /code/onnxruntime/include/onnxruntime/core/common/make_unique.h:28,
                 from /code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc:4:
/usr/include/c++/6/cstdlib:180:3: note: candidate: long long int std::abs(long long int)
   abs(long long __x) { return __builtin_llabs (__x); }
   ^~~
/usr/include/c++/6/cstdlib:172:3: note: candidate: long int std::abs(long int)
   abs(long __i) { return __builtin_labs(__i); }
   ^~~
CMakeFiles/onnxruntime_global_thread_pools_test.dir/build.make:75: recipe for target 'CMakeFiles/onnxruntime_global_thread_pools_test.dir/code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc.o' failed
make[2]: *** [CMakeFiles/onnxruntime_global_thread_pools_test.dir/code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc.o] Error 1
CMakeFiles/Makefile2:771: recipe for target 'CMakeFiles/onnxruntime_global_thread_pools_test.dir/all' failed
make[1]: *** [CMakeFiles/onnxruntime_global_thread_pools_test.dir/all] Error 2
make: *** [all] Error 2
Makefile:140: recipe for target 'all' failed
Traceback (most recent call last):
  File "/code/onnxruntime/tools/ci_build/build.py", line 1707, in <module>
    sys.exit(main())
  File "/code/onnxruntime/tools/ci_build/build.py", line 1662, in main
    build_targets(args, cmake_path, build_dir, configs, args.parallel)
  File "/code/onnxruntime/tools/ci_build/build.py", line 910, in build_targets
    run_subprocess(cmd_args, env=env)
  File "/code/onnxruntime/tools/ci_build/build.py", line 416, in run_subprocess
    completed_process = subprocess.run(
  File "/usr/local/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/local/bin/cmake', '--build', '/code/onnxruntime/build/Linux/MinSizeRel', '--config', 'MinSizeRel']' returned non-zero exit status 2.
The command '/bin/sh -c ./build.sh --use_openmp ${BUILDARGS} --build_shared_lib' returned a non-zero code: 1

Thanks @snnn for the suggestion. After changing the line, I have got a different meaningful error. Here it is,
/code/onnxruntime/onnxruntime/test/global_thread_pools/test_inference.cc:61:30: error: expected primary-expression before ‘>’ token
     ASSERT_TRUE(std::abs<OutT>(values_y[i] - f[i]) < 1e-6);

And after this, it is the usual make error.

--- a/onnxruntime/core/optimizer/utils.cc
+++ b/onnxruntime/core/optimizer/utils.cc
@@ -139,7 +139,7 @@ bool IsAttributeWithExpectedValue(const Node& node, const std::string& attr_name
 bool IsAttributeWithExpectedValue(const Node& node, const std::string& attr_name, float expected_value, float eps) {
   const auto* attr_proto = graph_utils::GetNodeAttribute(node, attr_name);
   if ((nullptr != attr_proto) && attr_proto->has_f()) {
-    return abs(attr_proto->f() - expected_value) < eps;
+    return std::abs(attr_proto->f() - expected_value) < eps;
   }
   return false;
 }

--- a/onnxruntime/test/global_thread_pools/test_inference.cc
+++ b/onnxruntime/test/global_thread_pools/test_inference.cc
@@ -57,7 +57,7 @@ static void RunSession(OrtAllocator& allocator, Ort::Session& session_object,
 
   OutT* f = output_tensor->GetTensorMutableData<OutT>();
   for (size_t i = 0; i != static_cast<size_t>(5); ++i) {
-    ASSERT_TRUE(std::abs(values_y[i] - f[i]) < 1e-6);
+    ASSERT_NEAR(values_y[i], f[i], 1e-6f);
   }
 }
 

